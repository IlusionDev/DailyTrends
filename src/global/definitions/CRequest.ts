import { Request } from "express-serve-static-core";
import { Logger } from "winston";

export default interface CRequest extends Request {
  logger: Logger;
}
