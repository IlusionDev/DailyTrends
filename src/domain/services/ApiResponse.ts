import ApiResponseDto from "@/domain/dto/ApiResponseDto";
import CRequest from "@/global/definitions/CRequest";
import { plainToClass } from "class-transformer";
import GenerateApiResponseOptions from "@/domain/services/definitions/GenerateApiResponseOptions";
import GenerateAndSendErrorReponseOptions from "@/domain/services/definitions/GenerateAndSendErrorReponseOptions";
import { Response } from "express";
import ErrorApi from "../../global/errors/ErrorApi";

export default class ApiResponseService {
  private req: CRequest;

  private res: Response;

  constructor(req: CRequest, res: Response) {
    this.req = req;
    this.res = res;
  }

  generateApiResponse<T>(
    data: T,
    options?: GenerateApiResponseOptions
  ): ApiResponseDto {
    const apiResponse: ApiResponseDto = new ApiResponseDto();
    if (data) {
      if (options?.sanitize) {
        apiResponse.data = plainToClass(options.sanitize, data);
      } else {
        apiResponse.data = data;
      }
    }
    apiResponse.error = options?.error ? options.error : false;
    options?.message ? (apiResponse.message = options?.message) : null;
    apiResponse.statusCode = options?.statusCode || 200;
    apiResponse.stack = options?.stack;

    return apiResponse;
  }

  generateApiErrorResponse<T>(
    options?: GenerateApiResponseOptions
  ): ApiResponseDto {
    const apiResponse = this.generateApiResponse(null, {
      error: true,
      statusCode: 500,
      ...options,
    });

    return apiResponse;
  }

  generateAndSendWithStatus(data: any, options?: GenerateApiResponseOptions) {
    const statusCode = options?.statusCode ? options.statusCode : 200;
    const response = this.res
      .status(statusCode)
      .send(this.generateApiResponse(data, options));

    return response;
  }

  generateAndSendErrorReponse(
    error: ErrorApi,
    options?: GenerateAndSendErrorReponseOptions
  ) {
    let statusCode: number = error?.statusCode
      ? error.statusCode
      : (500 as number);
    statusCode = options?.statusCode ? options.statusCode : statusCode;
    this.req.logger.error(
      error?.message || options?.message,
      error?.stack || ""
    );
    const stack = error.stack;
    return this.res.status(statusCode).send(
      this.generateApiResponse(null, {
        ...options,
        error: true,
        message: error?.message || options?.error?.message,
        statusCode,
        stack,
      })
    );
  }
}
