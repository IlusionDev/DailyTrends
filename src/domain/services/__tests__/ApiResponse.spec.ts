import ApiResponse from "@/domain/services/ApiResponse";
import CRequest from "@/global/definitions/CRequest";
import { Response } from "express-serve-static-core";

jest.mock("@/config/winston");

const req = {
  logger: {error: jest.fn() as any},
};

const res = {
  send: jest.fn() as any,
  status: jest.fn(() => res) as any,
};

const apiResponse = new ApiResponse(req as CRequest, res as Response);

describe("Check api Response service", () => {

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it("generates an api response with data", () => {
    const response = apiResponse.generateApiResponse({ name: "Entity" });

    expect(response).toMatchSnapshot()
  });

  it("generates an api response with data and options", () => {
    const response = apiResponse.generateApiResponse({ name: "Entity" }, {message: 'Message test', statusCode: 400, error: true});

    expect(response).toMatchSnapshot()
  });

   it("generates an error api response with default options", () => {
    const response = apiResponse.generateApiErrorResponse({message: 'Message test'});

    expect(response).toMatchSnapshot()
  });
   it("generates an error api response with custom statusCode", () => {
    const response = apiResponse.generateApiErrorResponse({message: 'Message test', statusCode: 400});

    expect(response).toMatchSnapshot()
  });
   it("generates an api response and sends it to client", () => {
    apiResponse.generateAndSendWithStatus({ name: "Entity" },{message: 'Message test', statusCode: 400});

     expect(res.status).toHaveBeenNthCalledWith(1, 400)
     expect(res.send).toHaveBeenNthCalledWith(1, {
       "data": {
        "name": "Entity"
       },
       "error": false,
       "message": "Message test",
       "statusCode": 400,
     })
  });
   it("generates an api response and sends it to client with default options", () => {
    apiResponse.generateAndSendWithStatus({ name: "Entity" });

     expect(res.status).toHaveBeenNthCalledWith(1, 200)
     expect(res.send).toHaveBeenNthCalledWith(1, {
       "data": {
        "name": "Entity"
       },
       "error": false,
       "message": undefined,
       "statusCode": 200,
     })
  });
   it("generates an error api response and sends it to client with default options", () => {
     apiResponse.generateAndSendErrorReponse(true, { error: { stack: 'stack example' } });

     expect(res.status).toHaveBeenNthCalledWith(1, 500)
     expect(res.send).toHaveBeenNthCalledWith(1, {
       "data": undefined,
       "error": true,
       "message": undefined,
       "statusCode": 500,
     })
  });
});
