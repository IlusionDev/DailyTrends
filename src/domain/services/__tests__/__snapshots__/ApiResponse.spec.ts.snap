// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check api Response service generates an api response with data 1`] = `
ApiResponseDto {
  "data": Object {
    "name": "Entity",
  },
  "error": false,
  "message": undefined,
  "s": undefined,
  "stack": undefined,
  "statusCode": 200,
}
`;

exports[`Check api Response service generates an api response with data and options 1`] = `
ApiResponseDto {
  "data": Object {
    "name": "Entity",
  },
  "error": true,
  "message": "Message test",
  "s": undefined,
  "stack": undefined,
  "statusCode": 400,
}
`;

exports[`Check api Response service generates an error api response with custom statusCode 1`] = `
ApiResponseDto {
  "data": undefined,
  "error": true,
  "message": "Message test",
  "s": undefined,
  "stack": undefined,
  "statusCode": 400,
}
`;

exports[`Check api Response service generates an error api response with default options 1`] = `
ApiResponseDto {
  "data": undefined,
  "error": true,
  "message": "Message test",
  "s": undefined,
  "stack": undefined,
  "statusCode": 500,
}
`;
